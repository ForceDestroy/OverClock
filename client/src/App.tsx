import { BrowserRouter, Route, Routes, useLocation } from 'react-router-dom';
import { Box, Grommet } from 'grommet';
import './App.css';

import Home from './pages/Home';
import NotFound from './pages/NotFound';
import Login from './pages/Login';
import EmployeeDashboard from './pages/EmployeeDashboard';
import EmployerDashboard from './pages/EmployerDashboard';
import NavigationTopBar from './components/NavigationTopBar';
import ProfilePage from './pages/ProfilePage';
import Timesheets from './pages/Timesheets';
import Schedule from './pages/Schedule';
import TimeOff from './pages/TimeOff';
import EmployerTimesheets from './pages/EmployerTimesheets';
import Payslip from './pages/Payslip';
import EmployerRequests from './pages/EmployerRequests';
import CompanyPolicy from './pages/CompanyPolicy';

export function App() {
  const { pathname } = useLocation();
  const toggleNav = pathname === '/Login' || pathname === '/';
  const availableRoutes = [
    '/employeedashboard',
    '/employeraccount',
    '/employerdashboard',
    '/employer/timesheets',
    '/employer/requests',
    '/application',
    '/applicationreferral',
    '/applicationmanagement',
    '/companydashboard',
    '/employeedashboard/companypolicy',
    '/profilepage',
    '/employeedashboard/timesheets',
    '/employeedashboard/schedule',
    '/employeedashboard/payslip',
    '/employeedashboard/timeoff',
  ];
  const errorRoute = availableRoutes.includes(pathname.toLowerCase());
  return (
    <div className="App">
      <Grommet full>
        {!toggleNav && errorRoute && <NavigationTopBar />}
        <Box direction="row" height={{ min: '90vh' }} background="light-1">
          <Routes>
            <Route path="/" element={<Home />} />
            <Route path="/Login" element={<Login />} />
            <Route path="/EmployeeDashboard" element={<EmployeeDashboard />} />
            <Route path="/EmployerDashboard" element={<EmployerDashboard />} />
            <Route path="/Employer/Requests" element={<EmployerRequests />} />
            <Route
              path="/Employer/Timesheets"
              element={<EmployerTimesheets />}
            />
            <Route
              path="/EmployeeDashboard/CompanyPolicy"
              element={<CompanyPolicy />}
            />
            <Route path="/ProfilePage" element={<ProfilePage />} />
            <Route
              path="/EmployeeDashboard/Timesheets"
              element={<Timesheets />}
            />
            <Route path="/EmployeeDashboard/Schedule" element={<Schedule />} />
            <Route path="/EmployeeDashboard/Payslip" element={<Payslip />} />
            <Route path="/EmployeeDashboard/TimeOff" element={<TimeOff />} />
            <Route path="*" element={<NotFound />} />
          </Routes>
        </Box>
      </Grommet>
    </div>
  );
}

export function WrappedApp() {
  return (
    <BrowserRouter>
      <App />
    </BrowserRouter>
  );
}
