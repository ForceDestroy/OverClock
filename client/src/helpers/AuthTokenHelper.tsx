/* eslint-disable consistent-return */
export default class AuthTokenHelper {
  static getFromLocalStorage = (): string | null => {
    try {
      if (localStorage.getItem('JWT_OC') === null) {
        throw ReferenceError();
      }
      const token = localStorage.getItem('JWT_OC');

      return token;
    } catch (error: unknown) {
      return null;
    }
  };

  static saveToLocalStorage = (token: string, timestamp?: string) => {
    try {
      localStorage.setItem('JWT_OC', token);
      if (timestamp) {
        localStorage.setItem('JWT_OC_TIMESTAMP', timestamp);
      } else {
        localStorage.setItem(
          'JWT_OC_TIMESTAMP',
          new Date().getTime().toString()
        );
      }
    } catch (error: unknown) {
      return null;
    }
  };

  static updateInLocalStorage = (token: string, timestamp?: string) => {
    try {
      if (localStorage.getItem('JWT_OC') === null) {
        throw ReferenceError();
      }
      localStorage.removeItem('JWT_OC');
      localStorage.setItem('JWT_OC', token);
      localStorage.removeItem('JWT_OC_TIMESTAMP');
      if (timestamp) {
        localStorage.setItem('JWT_OC_TIMESTAMP', timestamp);
      } else {
        localStorage.setItem(
          'JWT_OC_TIMESTAMP',
          new Date().getTime().toString()
        );
      }
    } catch (error: unknown) {
      return null;
    }
  };

  static removeFromLocalStorage = () => {
    try {
      if (localStorage.getItem('JWT_OC') === null) {
        throw ReferenceError();
      }
      localStorage.removeItem('JWT_OC');
      localStorage.removeItem('JWT_OC_TIMESTAMP');
      localStorage.removeItem('JWT_OC_NAME');
      localStorage.removeItem('JWT_OC_ACCESS_LEVEL');
      localStorage.removeItem('JWT_OC_USER_ID');
    } catch (error: unknown) {
      return null;
    }
    return null;
  };

  static validateToken = (): boolean => {
    try {
      if (localStorage.getItem('JWT_OC') === null) {
        throw ReferenceError();
      }
      const tokenTimestamp = localStorage.getItem('JWT_OC_TIMESTAMP');
      const timestamp = Date.parse(tokenTimestamp || '');

      if (timestamp + 30 * 60 * 1000 < new Date().getTime()) {
        this.removeFromLocalStorage();
        return false;
      }

      return true;
    } catch (error: unknown) {
      return false;
    }
  };

  static setName = (name: string) => {
    try {
      if (localStorage.getItem('JWT_OC') === null) {
        throw ReferenceError();
      }
      localStorage.setItem('JWT_OC_NAME', name);
    } catch (error: unknown) {
      return null;
    }
  };

  static getName = (): string | null => {
    try {
      if (localStorage.getItem('JWT_OC') === null) {
        throw ReferenceError();
      }
      const name = localStorage.getItem('JWT_OC_NAME');

      return name;
    } catch (error: unknown) {
      return null;
    }
  };

  static setAccessLevel = (accessLevel: string) => {
    try {
      if (localStorage.getItem('JWT_OC') === null) {
        throw ReferenceError();
      }
      localStorage.setItem('JWT_OC_ACCESS_LEVEL', accessLevel);
    } catch (error: unknown) {
      return null;
    }
  };

  static getAccessLevel = (): string | null => {
    try {
      if (localStorage.getItem('JWT_OC') === null) {
        throw ReferenceError();
      }
      const accessLevel = localStorage.getItem('JWT_OC_ACCESS_LEVEL');

      return accessLevel;
    } catch (error: unknown) {
      return null;
    }
  };

  static setUserId = (userId: string) => {
    try {
      if (localStorage.getItem('JWT_OC') === null) {
        throw ReferenceError();
      }
      localStorage.setItem('JWT_OC_USER_ID', userId);
    } catch (error: unknown) {
      return null;
    }
  };

  static getUserId = (): string | null => {
    try {
      if (localStorage.getItem('JWT_OC') === null) {
        throw ReferenceError();
      }
      const userId = localStorage.getItem('JWT_OC_USER_ID');

      return userId;
    } catch (error: unknown) {
      return null;
    }
  };
}
