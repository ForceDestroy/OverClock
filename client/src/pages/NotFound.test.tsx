import { render, screen, waitFor, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import { describe, it, vi } from 'vitest';
import AuthTokenHelper from '../helpers/AuthTokenHelper';
import { App } from '../App';

vi.mock('axios');
vi.mock('../helpers/AuthTokenHelper');

describe('NotFound', () => {
  afterEach(() => {
    vi.clearAllMocks();
    vi.resetAllMocks();
  });
  it('Renders Not Found to match snapshot', () => {
    // Arrange
    const snap = render(
      <MemoryRouter initialEntries={['/something']}>
        <App />
      </MemoryRouter>
    );

    // Assert
    expect(snap).toMatchSnapshot();
  });
  it('Renders Not Found on a non-existent page', () => {
    // Arrange
    render(
      <MemoryRouter initialEntries={['/something']}>
        <App />
      </MemoryRouter>
    );

    // Act
    // No Act for this Test

    // Assert
    expect(screen.getByText('This page does not exist')).toBeInTheDocument();
  });
  it('Renders Not Found on a non-existent page and navigate back home', async () => {
    // Arrange
    render(
      <MemoryRouter initialEntries={['/something']}>
        <App />
      </MemoryRouter>
    );
    const user = userEvent.setup();

    // Act
    const homeButton = screen.getByTestId('goHomeButton');
    await user.click(homeButton);

    // Assert
    await waitFor(() => {
      const links: HTMLAnchorElement[] = screen.getAllByRole('link');
      expect(links[0].href).toContain('/');
    });
  });
  it('Renders Not Found on a non-existent page and navigate to EmployeeDashboard if logged in', async () => {
    // Mock
    (AuthTokenHelper.getFromLocalStorage as jest.Mock).mockReturnValue(
      'VerySecureToken1234567890'
    );

    // Arrange
    // const user = userEvent.setup();

    render(
      <MemoryRouter initialEntries={['/something']}>
        <App />
      </MemoryRouter>
    );

    // Act
    const homeButton = screen.getByTestId('goHomeButton') as HTMLAnchorElement;

    // Assert
    await waitFor(() => {
      expect(homeButton.href).toContain('/EmployeeDashboard');
    });
  });
});
