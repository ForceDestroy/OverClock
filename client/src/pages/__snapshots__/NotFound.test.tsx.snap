// Vitest Snapshot v1

exports[`NotFound > Renders Not Found to match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <div
          class="StyledGrommet-sc-19lkkz7-0 jGVJhU"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 eETalo"
          >
            <div
              class="StyledGrommet-sc-19lkkz7-0 jGVJhU"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 hwtiCx"
              >
                <span
                  class="StyledText-sc-1sadyjn-0 jEBSos"
                >
                  This page does not exist
                </span>
                <div
                  class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ceHRnz"
                />
                <div
                  class="StyledBox-sc-13pk1d4-0 iNDmPr"
                >
                  <img
                    alt="404"
                    src="/src/assets/gif/cry.gif"
                  />
                </div>
                <div
                  class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ceHRnz"
                />
                <div
                  class="StyledBox-sc-13pk1d4-0 jzOfgr"
                >
                  <a
                    class="StyledButton-sc-323bzc-0 byZKYO"
                    data-testid="goHomeButton"
                    href="/"
                  >
                    Go Home
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <div
        class="StyledGrommet-sc-19lkkz7-0 jGVJhU"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 eETalo"
        >
          <div
            class="StyledGrommet-sc-19lkkz7-0 jGVJhU"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 hwtiCx"
            >
              <span
                class="StyledText-sc-1sadyjn-0 jEBSos"
              >
                This page does not exist
              </span>
              <div
                class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ceHRnz"
              />
              <div
                class="StyledBox-sc-13pk1d4-0 iNDmPr"
              >
                <img
                  alt="404"
                  src="/src/assets/gif/cry.gif"
                />
              </div>
              <div
                class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ceHRnz"
              />
              <div
                class="StyledBox-sc-13pk1d4-0 jzOfgr"
              >
                <a
                  class="StyledButton-sc-323bzc-0 byZKYO"
                  data-testid="goHomeButton"
                  href="/"
                >
                  Go Home
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
